planner/core/task.go

```diff
@@ -1483,6 +1572,71 @@ func (p *PhysicalHashAgg) cpuCostDivisor(hasDistinct bool) (float64, float64) {
 	return math.Min(float64(finalCon), float64(partialCon)), float64(finalCon + partialCon)
 }
 
+func (p *PhysicalHashAgg) attach2TaskForMpp(tasks ...task) task {
+	t := tasks[0].copy()
+	mpp, ok := t.(*mppTask)
+	if !ok {
+		return invalidTask
+	}
+	inputRows := mpp.count()
+	switch p.MppRunMode {
+	case Mpp1Phase:
+		/// 1-phase agg: when the partition columns can be satisfied, where the plan does not need to enforce Exchange
+		/// only push down the original agg
+		prop := &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.HashType, PartitionCols: mpp.hashCols}
+		proj := p.convertAvgForMPP(prop)
+		p.self.SetChildren(mpp.p)
+		mpp.p = p.self
+		if proj != nil {
+			proj.SetChildren(mpp.p)
+			mpp.p = proj
+		}
+		mpp.addCost(p.GetCost(inputRows, false))
+		return mpp
+	case Mpp2Phase:
+		/// 2-phase agg: partial + final agg for hash partition
+		if len(p.PartitionCols) == 0 {
+			return invalidTask
+		}
+		prop := &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.HashType, PartitionCols: p.PartitionCols}
+		// if mpp does not need to enforce exchange, i.e., the child is properly partitioned, then this 2-phase agg is invalid
+		if !mpp.needEnforce(prop) {
+			return invalidTask
+		}
+		proj := p.convertAvgForMPP(prop)
+		partialAgg, finalAgg := p.newPartialAggregate(kv.TiFlash)
+		if partialAgg == nil {
+			return invalidTask
+		}
+		partialAgg.SetChildren(mpp.p)
+		mpp.p = partialAgg
+		newMpp := mpp.enforceExchangerImpl(prop)
+		finalAgg.SetChildren(newMpp.p)
+		newMpp.p = finalAgg
+		if proj != nil {
+			proj.SetChildren(newMpp.p)
+			newMpp.p = proj
+		}
+		// TODO: how to set 2-phase cost?
```

Comment:
Since the optimizer has no physical information about data and cluster instances like how many TiFlash nodes will be involved in this plan, it is impossible to get an accurate cost here.

Reply:
yeah, this is a hard problem.